src/tag_storage.rs

```diff
@@ -0,0 +1,245 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+///! This is the database that holds tags associated to various objects.
+///! It provides an api to manage Id <-> tags relationships.
+///! All users share the same tags for objects.
+
+use rusqlite::{ Connection, Result };
+use std::path::PathBuf;
+use util::{ Id, TagId };
+
+fn escape<T>(string: &Id<T>) -> String {
+    // http://www.sqlite.org/faq.html#q14
+    format!("{}", string).replace("'", "''")
+}
+
+/// Creates a unique key for a (id, tag) tuple.
+/// `SQlite` integers are i64 so we turn the hashed u64 into a String...
+fn create_key<T>(id: &Id<T>, tag: &Id<TagId>) -> String {
+    use std::hash::{ Hash, Hasher, SipHasher };
+
+    let mut hasher = SipHasher::new();
+    id.hash(&mut hasher);
+    tag.hash(&mut hasher);
+    format!("{}", hasher.finish())
+}
+
+/// A lighweight struct to manage the database. Creating these objects is very cheap because the
+/// underlying database is created lazily when we need it.
+pub struct TagStorage {
+    db: Option<Connection>,
+    path: PathBuf,
+}
+
+impl TagStorage {
+    pub fn new(path: &PathBuf) -> Self {
+        TagStorage {
+            db: None,
+            path: path.clone()
+        }
+    }
+
+    // Ensures that we have a database ready. If we fail to open or create the database,
+    // this will panic.
```

Comment:
Not a big fan of panics. Well, let's start with this and look for a better way to handle that.